compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/mm_npu4.cc

```diff
@@ -0,0 +1,193 @@
+#define REL_READ 1
+
+#ifndef __chess__
+#include "aiebase_chess.h"
+#include "aiebase_typedefs.h"
+#endif
+
+#include <aie_api/aie.hpp>
+
+template<typename T, int M, int N, int val>
+void zero_vectorized(T *__restrict pC, unsigned offsetC)
+{
+  T *__restrict pC1 = pC + offsetC;
+  for (unsigned r = 0; r < M; r += 1) {
+    for (unsigned c = 0; c < N; c += 1) {
+      unsigned o0 = N * r + c;
+      pC1[o0] = T(0);
+    }
+  }
+}
+
+
+template <typename T_in> aie::accum<accfloat, 64> load_v64(const T_in *__restrict p) {
+  aie::vector<T_in, 32> v0 = aie::load_v<32>(p);
+  aie::accum<accfloat, 32> accum0 = aie::accum<accfloat, 32>(v0);
+  aie::vector<T_in, 32> v1 = aie::load_v<32>(p + 32);
+  aie::accum<accfloat, 32> accum1 = aie::accum<accfloat, 32>(v1);
+  aie::accum<accfloat, 64>  accum = aie::concat(accum0, accum1);
+  return accum;
+}
+
+template <typename T_in> aie::accum<accfloat, 64> load_v64_T(const T_in *__restrict p) {
+  aie::vector<T_in, 32> v0 = aie::load_v<32>(p);
+  aie::vector<T_in, 32> v1 = aie::load_v<32>(p + 32);
+  aie::vector<T_in, 32> v0Shuffled = ::shuffle(v0, 29);//shuffle_T16_4x8
+  aie::vector<T_in, 32> v1Shuffled = ::shuffle(v1, 29);//shuffle_T16_4x8
+  aie::vector<T_in, 32> vShuffledLo = ::shuffle(v0Shuffled, v1Shuffled, 14);//shuffle_T64_2x8_lo
+  aie::vector<T_in, 32> vShuffledHi = ::shuffle(v0Shuffled, v1Shuffled, 15);//shuffle_T64_2x8_hi
+  aie::accum<accfloat, 32> accum0 = aie::accum<accfloat, 32>(vShuffledLo);
+  aie::accum<accfloat, 32> accum1 = aie::accum<accfloat, 32>(vShuffledHi);
+  aie::accum<accfloat, 64>  accum = aie::concat(accum0, accum1);
+  return accum;
+}
+
+template<unsigned rowA, unsigned colA, unsigned colB, unsigned L0_M, unsigned L0_K, unsigned L0_N>
+void matmul_vectorized_bf16_f32(const bfloat16 * __restrict pA, unsigned offsetA, const bfloat16 * __restrict pB, unsigned offsetB, float * __restrict pC, unsigned offsetC)
+{
+  const unsigned size_A = L0_M * L0_K;
+  const unsigned size_B = L0_K * L0_N;
+  const unsigned size_C = L0_M * L0_N;
+  using MMUL = aie::detail::mmul_bfp16_bfp16<L0_M, L0_K, L0_N, bfp16ebs8, bfp16ebs8, 32>;
+  
+  v32accfloat * restrict pOut = (v32accfloat *) (pC + offsetC);
+
+  for (unsigned z = 0; z < rowA; z += 2)
+    chess_loop_range(4, ) {
+      float *__restrict pC1 = pC + offsetC + (z)*size_C;
+      float *__restrict pC2 = pC + offsetC + ((z + 1)) * size_C;
+
+
+      for (unsigned j = 0; j < colB; j += 2)
+        chess_prepare_for_pipelining chess_loop_range(4, ) {
+          const bfloat16 *__restrict pA1 = pA + offsetA + (z)*size_A;
+          const bfloat16 *__restrict pA2 = pA + offsetA + ((z + 1)) * size_A;
+          const bfloat16 *__restrict pB1 = pB + offsetB + (j)*colA*size_B;
+          const bfloat16 *__restrict pB2 = pB + offsetB + ((j + 1))*colA * size_B;
+
+          aie::accum<accfloat, size_A>  accumA0 = load_v64<bfloat16>(pA1);
+          aie::bfp_vector<bfp16ebs8, size_A> A0 = accumA0.template to_vector<bfp16ebs8>();
```

Comment:
I don't think we'll ever want to do this in 'production' code, we'll be moving bfp16 in from L2 not casting to it on the core. For this reason, I wonder if it would be better use of time (and possibly simpler) to work on i8->i32 ukernels.

Reply:
It's the best way to support bfloat16 on strix, even though it's exactly the same accuracy. I think this is quite useful for now until bfp16 will be supported, which can take a while.

Reply:
But we should add an i8 kernel as well.